File: src/round.cpp
#include "round.h"

RetriableMethodResponse Round::attemptProgress() {
    // State: Soliciting pirate selection
    if (pirates.size() < Game::instance().getPlayers().size()) {
        // solicit choices from each player
        if (pirates.size() == Game::instance().getPlayers().size()) {
            // State: All players have submitted a pirate, sort them and set dayIterator
            sort(pirates.begin(), pirates.end());
            dayIterator = pirates.begin();
            return RetriableMethodResponse::MadeProgress;
        } else {
            return RetriableMethodResponse::PendingInput;
        }
    }
    
    // State: Iterating through day actions of pirates
    if (dayIterator != pirates.end()) {
        RetriableMethodResponse response = dayIterator->dayAction();
        if (response == RetriableMethodResponse::Complete) {
            dayIterator++;
            // State: Finished iterating over day actions, reverse vector and set duskIterator
            if (dayIterator == pirates.end()) {
                reverse(pirates.begin(), pirates.end());
                duskIterator = pirates.begin();
                return RetriableMethodResponse::MadeProgress;
            }
        } else {
            return response;
        }
    }
    
    // State: Iterating through dusk actions of pirates
    if (duskIterator != pirates.end()) {
        RetriableMethodResponse response = duskIterator->duskAction();
        if (response == RetriableMethodResponse::Complete) {
            duskIterator++;
        } else {
            return response;
        }
    }

    // State: Finished iterating over dusk actions, return true; round is done
    if (duskIterator == pirates.end()) {
        return RetriableMethodResponse::Complete;
    }
    
    return RetriableMethodResponse::MadeProgress;
}

void Round::endRound() {
    // Move surviving pirates to den
    for (Pirate &p : pirates) {
        
        // TOOD: fix me!
        Game::instance().getPlayer(p.getOwnerId()).getDen();
    }
    
    // Do night time actions
    for (Player &p : Game::instance().getPlayers()) {
        p.doNightActions();
    }
}

File: src/round.scala
import scala.collection.mutable.ArrayBuffer

class Round(val booty : Seq[Booty.Value]) {
    
    var state : RoundState.Value = RoundState.SolicitPirates
    val pirates : Seq[Pirate] = ArrayBuffer[Pirate]()
    
    def makeProgress() : RetriableMethodResponse.Value = {
        if (state == RoundState.SolicitPirates) {
            var response = solicitPirates()
            if (response != RetriableMethodResponse.Complete) {
                return response
            }
        } else if (state == RoundState.DayActions) { 
            // order pirates in ascending order
        } else if (state == RoundState.DuskActions) {
            // sort pirates in descending order
        } else if (state == RoundState.NightActions) {
            //
        }
        
        return RetriableMethodResponse.Complete;   
    }
  
    // only repeat actions which can be repeated
    // the only actions which require repeats are those which have player input
    // each code block should only contain the repeatable aspects
    // only code repeats for situations which can have them (inputs)

    private def solicitPirates() : RetriableMethodResponse.Value = {
        var pendingInput : Boolean = false
        for (p <- PlayerManager.players) {
            val request = InputManager.postAndGetInputRequest(
                p.playerId,
                InputRequestType.PlayPirateFromHand,
                InputManager.getPlayerHandFromPlayer(p))
            if (!request.answered) {
                pendingInput = true
            }
        }
        if (pendingInput) {
            println("Waiting for players to select pirates")
            return RetriableMethodResponse.PendingInput
        } else {
            // load the pirates into the round
            state = RoundState.DayActions;
            return RetriableMethodResponse.Complete
        }
        
    }
    
    private def performDayActions() : RetriableMethodResponse.Value = {
        // return RetriableMethodResponse.PendingInput
        state = RoundState.DuskActions
        return RetriableMethodResponse.Complete
    }
    
    private def performDuskActions() : RetriableMethodResponse.Value = {
        // return RetriableMethodResponse.PendingInput
        state = RoundState.NightActions
        return RetriableMethodResponse.Complete
    }
    
    private def performNightActions() : RetriableMethodResponse.Value = {
        return RetriableMethodResponse.Complete
    }
    
    def endRound() = {
        
    }
}

object RoundState extends Enumeration {
    type RoundState = Value
    val SolicitPirates, DayActions, DuskActions, NightActions = Value
}
